File: kt/Minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/Parser.kt
package dev.phytolizer.minsk.analysis.syntax

class Parser(text: String) {
    private val _tokens: List<SyntaxToken>
    private var _position = 0
    private val _diagnostics = mutableListOf<String>()
    val diagnostics: List<String>
        get() = _diagnostics

    init {
        val tempTokens = mutableListOf<SyntaxToken>()
        val lexer = Lexer(text)
        for (token in lexer) {
            if (token.kind != SyntaxKind.BadToken && token.kind != SyntaxKind.WhitespaceToken) {
                tempTokens.add(token)
            }
        }
        tempTokens.add(SyntaxToken(SyntaxKind.EndOfFileToken, tempTokens.lastOrNull()?.position ?: 0, "", null))
        _diagnostics.addAll(lexer.diagnostics)
        _tokens = tempTokens
    }

    private fun peek(offset: Int): SyntaxToken {
        val index = _position + offset
        if (index >= _tokens.size) {
            return _tokens.last()
        }

        return _tokens[index]
    }

    private val current
        get() = peek(0)

    private fun nextToken(): SyntaxToken {
        val curr = current
        _position += 1
        return curr
    }

    private fun matchToken(kind: SyntaxKind): SyntaxToken {
        if (current.kind == kind) {
            return nextToken()
        }

        _diagnostics.add("Expected next token to be $kind, got ${current.kind} instead")
        return SyntaxToken(kind, current.position, "", null)
    }

    fun parse(): SyntaxTree {
        return SyntaxTree(parseExpression(), matchToken(SyntaxKind.EndOfFileToken), diagnostics)
    }

    private fun parseExpression(): ExpressionSyntax {
        return parseTerm()
    }

    private fun parseTerm(): ExpressionSyntax {
        var left = parseFactor()

        while (current.kind == SyntaxKind.PlusToken || current.kind == SyntaxKind.MinusToken) {
            val operatorToken = nextToken()
            val right = parseFactor()
            left = BinaryExpressionSyntax(left, operatorToken, right)
        }

        return left
    }

    private fun parseFactor(): ExpressionSyntax {
        var left = parsePrimaryExpression()

        while (current.kind == SyntaxKind.StarToken || current.kind == SyntaxKind.SlashToken) {
            val operatorToken = nextToken()
            val right = parsePrimaryExpression()
            left = BinaryExpressionSyntax(left, operatorToken, right)
        }

        return left
    }

    private fun parsePrimaryExpression(): ExpressionSyntax {
        return if (current.kind == SyntaxKind.OpenParenthesisToken) {
            val openParenthesisToken = matchToken(SyntaxKind.OpenParenthesisToken)
            val expression = parseExpression()
            val closeParenthesisToken = matchToken(SyntaxKind.CloseParenthesisToken)
            ParenthesizedExpressionSyntax(openParenthesisToken, expression, closeParenthesisToken)
        } else {
            val numberToken = matchToken(SyntaxKind.NumberToken)
            LiteralExpressionSyntax(numberToken)
        }
    }
}

File: py/minsk/analysis/syntax/parser.py
from minsk.analysis.syntax.expression import ExpressionSyntax
from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.lexer import Lexer
from minsk.analysis.syntax.token import SyntaxToken


class Parser:
    _tokens: tuple[SyntaxToken, ...]
    _position: int

    def __init__(self, text: str):
        lexer = Lexer(text)
        tokens = list(
            filter(
                lambda tok: tok.kind != SyntaxKind.BadToken and
                            tok.kind != SyntaxKind.WhitespaceToken,
                lexer
            )
        )
        tokens.append(SyntaxToken(SyntaxKind.EndOfFileToken, 0, "", None))
        self._tokens = tuple(tokens)
        self._position = 0

    def _peek(self, offset: int) -> SyntaxToken:
        index = self._position + offset
        if index >= len(self._tokens):
            return self._tokens[-1]
        return self._tokens[index]

    @property
    def _current(self) -> SyntaxToken:
        return self._peek(0)

    def parse(self) -> ExpressionSyntax:
        pass
